### ISSUE TITLE
Unable to easily select screen option from home screen with mobile VO

### ISSUE THREAD

**Issue by KatieWoe** [phetsims/ratio-and-proportion#311]
**Device**
iPad 6th Gen
**OS**
iPadOS 14.3
**Browser**
Safari
**Problem Description**
For https://github.com/phetsims/QA/issues/582
It is possible for the list containing the two screen option on the home screen to seemingly disappear. Swiping left/right to read through or swiping two fingers up will skip over the options, sometimes with blank options containing no words, so that you go from the brief intro to the mute button. If you click the buttons, they seem to come back, but otherwise you need to refresh the page.
Steps to reproduce
1. Refresh the page
2. Swipe two fingers up to start reading from the top
3. Listen until the end
4. Swipe back to try and find the screen options
5. Swipe two fingers up to start from the top again and listen to it skip these options

If you swipe to the side from the start or press the buttons directly the bug does not seem to happen or is canceled

**Visuals**
https://drive.google.com/file/d/1oLW_gXG9Rc4XC7R7_hfKi32Ktb3nNQMy/view?usp=sharing

**Comment by KatieWoe** [comment]
Non buggy, but similar behavior that I saw: on the first screen, if you read through the challenge options and go past them, the box closes. If you then swipe back it reads through a few blank options where the challenge list was. You can then open the list again and and read through the options again.
https://drive.google.com/file/d/1zD47i8-0hpg53FvOJwPgn_A5lbGczhig/view?usp=sharing

**Comment by zepumph** [comment]
I feel a bit like this goes beyond my knowledge about what is correct for screen reasers. @terracoda can you let me know what you think this is a bug to you?

**Comment by terracoda** [comment]
Hmm, it looks like the buttons disappear from the PDOM.
I'll have to try and reproduce on my devices.

**Comment by zepumph** [comment]
Perhaps this has to do with https://github.com/phetsims/scenery/issues/1094 then.

**Comment by terracoda** [comment]
@zepumph, it might be related to phetsims/scenery#1094 but I need to check on my 2 iphones. I have different iOS's on them, so might be able to verify.

Just need to figure out how to log into phettest from my iphone.

**Comment by terracoda** [comment]
I logged into phettest :-) on bayes.

**Comment by terracoda** [comment]
Hmm, the sim doesn't load from phettest.
@zepumph, when you get a chance can you make me a dev link.
In the mean time I will check dev 78 on both devices.

**Comment by terracoda** [comment]
Using VO and Safari on iPhone 8 with iOS 13.5.1, I do not have any trouble accessing the screen buttons on the Home page.

**Comment by zepumph** [comment]
>Using VO and Safari on iPhone 8 with iOS 13.5.1, I do not have any trouble accessing the screen buttons on the Home page.

Does this mean that you do have trouble on iOS 14?

**Comment by terracoda** [comment]
I can access the sim screen buttons easily on iPhone 7 with iOS 14.2

However, when I do exactly as @KatieWoe described, i.e., read from top with a 2 finger swipe up, then try to read backward with the left single-finger swipes, I can't get on the buttons.

If I read again from top with a 2-finger swipe up sometimes I get the buttons on a auto-re-read, and sometimes I don't.

There is something very buggy going on.

I am not sure, but could it have something to do with the fact that the buttons are kind of weird by nature. They kind of have a two-step activation process, first they get big, then you need to activate them to go tot he screen?

It seems if the second screen button is visually big (with no pink high light), and I try to read from top, or read with left to right swipes, it seems like the screen buttons are not there.

If the screen 1 button is visually big, reading from the top reads through the buttons, and reading with left and right swipes reads through the buttons.

The buttons seem to get skipped only when the second screen button is left visually big and I I try to go back or try to read from top. 

Question: 
What happens to the list of buttons in the PDOM when the screen 2 button is visually big?

Note that it is always left visually big after a read through, so it is important to know what's happening to the list of buttons in the PDOM in each phase of the button.

**Comment by terracoda** [comment]
@zepumph, re-assigning to you to answer my question about the PDOM.

**Comment by terracoda** [comment]
@zepumph, could this issue be related to #256?

**Comment by zepumph** [comment]
Things to think about in my investigation:
* Is this caused by the list structure?
* Is this caused by how we swap PDOM content when changing from a big button to a small button. Likely the last time that the HomeScreen had PDOM work on it was when we had two complete instances of a screen button for each sized button.

**Comment by zepumph** [comment]
Here is how I consistently reproduce this problem:

1. Load the sim on iOS.
2. Physically touch the PhET Menu (one of the few elements on the home screen still being positioned in the PDOM).
3. Swipe left back until the second, create screen is highlighted.
4. swipe two fingers up to read from the beginning, during this read-through, it will skip the second screen.


I am able to consistently find this issue fixed when I comment out this line https://github.com/phetsims/joist/blob/934a21b71a5b85379d2a0e56201cd4933a058f34/js/HomeScreenButton.js#L185

```js
this.addInputListener( { focus: event => { !isSelectedProperty.value && fireListener.fire( event ); } } );
```

My first guess as to why this is causing a problem, is because the fireListener in this case changes the `isSelectedProperty`, and I would assume that this listener effects the PDOM in some way that "interrupts" the VO read in some way. I don't see anything that would make me think that though:

```js
    // update pieces that change when the button is selected or unselected
    isSelectedProperty.link( isSelected => {
      const data = isSelected ? settings.large : settings.small;

      // apply settings for the current size
      nodeContainer.children = data.node;
      text.font = data.font;
      text.maxWidth = nodeContainer.width;
      setOpacityAndFill();
      this.setSpacing( data.spacing );

      // update the VBox layout after changing the size of its children
      this.updateLayout();
    } );
```
(I looked into this.updateLayout and it doesn't change the children order from what I see).

I'll keep investigating.

**Comment by zepumph** [comment]
The issue isn't the logic in HomeScreenButton, but instead the re-layout in HomeScreenView when selectedScreenProperty is changed:

https://github.com/phetsims/joist/blob/1acce4c15cc0158ba78dbcada5fe9760454a3fa4/js/HomeScreenView.js#L167

```js
     const updateIconsLayout = () => {
      const selectedScreen = model.selectedScreenProperty.value;

      // remove previous layout of icons
      if ( hBox ) {
        hBox.removeAllChildren(); // because icons have reference to hBox (their parent)
        iconsParentNode.removeChild( hBox );
      }

      // add new layout of icons
      const icons = _.map( screenElements, screenElement => {

        // check for the current screen
        if ( screenElement.screen === selectedScreen ) {
          this.highlightedScreenButton = screenElement.button;
        }
        return screenElement.button;
      } );

      hBox = new HBox( {
        spacing: spacing,
        children: icons,
        align: 'top',
        resize: false,
        maxWidth: this.layoutBounds.width - 118
      } );
      iconsParentNode.addChild( hBox );

      // position the icons
      iconsParentNode.centerX = this.layoutBounds.width / 2;
      iconsParentNode.top = this.layoutBounds.height / 3 + 20;
    };
```



It is too late in the day for me to try to figure out how to proceed. Do I adapt the home screen so that the scene graph isn't getting completely rewritten as we cycle through screen buttons? It seems ideal, but I would want to discuss this with past contributors of the home screen code. I think my first pick would be @chrisklus. To sum up the issue, VoiceOver while using iOS has a bug in its behavior and ability to read the home screen buttons. This is because when you focus an unselected button, the view will tear down and recreate the hbox of buttons. By removing this content from the PDOM and readding it, we lose one of the buttons from the PDOM completely in some cases. 

I will pick this back up on Monday.

**Comment by zepumph** [comment]
This has been fixed by the first commit over in https://github.com/phetsims/joist/issues/683. @terracoda will you please confirm solved.

**Comment by zepumph** [comment]
Sorry, I didn't check who opened the issue. @KatieWoe will you please confirm fixed on master.

**Comment by zepumph** [comment]
You can also test with https://phet-dev.colorado.edu/html/ratio-and-proportion/1.0.0-dev.79/phet/

**Comment by KatieWoe** [comment]
Checked on master and this seems fixed there

**Comment by terracoda** [comment]
Awesome, I'll give it a spin, too.

**Comment by terracoda** [comment]
Works perfectly with iOS 14.3 using VO on iPhone 7.

**Comment by terracoda** [comment]
Nice work @zepumph and @chrisklus! I think this issue can be closed.
